### DISCUSSION TITLE
Do delayed messages stop actors exiting?

### DISCUSSION THREAD

**Discussion by al3ph** [actor-framework/actor-framework#1209] [Q&A]
Not sure if this is a bug or a feature.. 

In my test bed an actor sends itself a heartbeat every 10 seconds, (to keep a connection alive), unfortunately this stops the actor exiting when the testbed exits.

Not sure how to handle this, I'm guessing the delayed msg in the queue is acting as a reference to the actor, which stops it being able to be garbage collected, or possibly the fact that the queue isn't empty?

Is there any method of making this work ? Do I manually have to kill the actor ? Though I assume it'd still not die until the pending msg is processed.

Any pointers would be very much appreciated 

cheer!

**Comment by Neverlord** [comment]
> Not sure if this is a bug or a feature..

A feature. 🙂

There are a couple things at play here:
- On shutdown, i.e., in the destructor of `actor_system`, CAF waits for all actors to terminate.
- An actor implicitly terminates when it becomes unreachable.

So by scheduling a delayed message. there is at least one reference to the actor until timing out. Then, the actor processes the message. If your system shuts down afterwards, that means that message had the last strong reference to the actor and it terminates because it became unreachable. If that has been the last actor, then the destructor of `actor_system` goes on to shut everything down.

> Is there any method of making this work ? Do I manually have to kill the actor ?

Killing the actor manually should give you the desired behavior. There is still a strong reference to it remaining (the scheduled message), but the system is going to shut down regardless if no actors are left running. The actor system should cancel all pending timeouts in this case.

**Comment by al3ph** [comment]
Is there a way of manually triggering the cancelling of delayed messages ? Might be a handy way of triggering the required behaviour, without manually killing any actors waiting on delayed msgs.

**Comment by Neverlord** [comment]
There's no API for cancelling asynchronous message individually. However, you could cancel all scheduled messages: `sys.clock().cancel_all()`: https://github.com/actor-framework/actor-framework/blob/master/libcaf_core/caf/actor_clock.hpp#L75.

Be cautious with this. Using it as part of your shutdown should be ok, but cancelling messages of other actors can easily wreak havoc. 😉

**Comment by al3ph** [comment]
Ah excellent, that exactly what I was after ! Cheers!
